src/Workspaces/Core/Portable/Workspace/FileTextLoader.cs

```diff
@@ -58,18 +60,39 @@ public FileTextLoader(string path, Encoding? defaultEncoding)
 
         internal sealed override string FilePath => Path;
 
-        protected virtual SourceText CreateText(Stream stream, Workspace workspace)
-        {
-            var factory = workspace.Services.GetRequiredService<ITextFactoryService>();
-            return factory.CreateText(stream, DefaultEncoding);
-        }
+        internal override bool CanReloadText => !IsObsoleteCreateTextOverridden;
+
+#pragma warning disable CS0618 // Type or member is obsolete
+        private bool IsObsoleteCreateTextOverridden
+            => s_isObsoleteCreateTextOverriden.GetValue(
+                GetType(), _ => new StrongBox<bool>(new Func<Stream, Workspace, SourceText>(CreateText).Method.DeclaringType != typeof(FileTextLoader))).Value;
+#pragma warning restore
+
+        /// <summary>
+        /// Creates <see cref="SourceText"/> from <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">Stream.</param>
+        /// <param name="workspace">Obsolete. Null.</param>
+        [Obsolete("Use CreateText(Stream, CancellationToken)")]
+        protected virtual SourceText CreateText(Stream stream, Workspace? workspace)
+            => EncodedStringText.Create(stream, DefaultEncoding, checksumAlgorithm: SourceHashAlgorithm.Sha256);
+
+        /// <summary>
+        /// Creates <see cref="SourceText"/> from <see cref="Stream"/>.
+        /// </summary>
+#pragma warning disable CS0618 // Type or member is obsolete
+        protected virtual SourceText CreateText(Stream stream, LoadTextOptions options, CancellationToken cancellationToken)
+            => IsObsoleteCreateTextOverridden ?
+                CreateText(stream, workspace: null) :
```

Comment:
>CreateText(stream, workspace: null) [](http://example.com/codeflow?start=16&length=35)

It is not clear to me why ignoring ```options.ChecksumAlgorithm``` is the right thing to do for this code path.

Reply:
I'm not sure what else could we do here. The legacy method does not take the options. 
`CanReloadText` returns false if this code path is taken which signals that the option is ignored.

Reply:
> I'm not sure what else could we do here. The legacy method does not take the options. 

I do not remember discussing this specific fall back at API review. While this type of delegation feels reasonable to me for a scenario when a parameter is getting removed and we pass some default value to a legacy method, I am not comfortable with this behavior for a scenario when a new parameter is getting added. Consumer's expectation will be that the parameter is not ignored.

Reply:
What fallback do you suggest that doesn't break existing code entirely? This approach keeps existing implementations working in the mode that they ignore the options (which is signaled by CanReloadText property). New implementations have the options available and should apply them.

Reply:
And yes, this wasn't discussed before. The options parameter is new. We discovered after the API discussion that passing it in is beneficial for the rest of the solution implementation.

Reply:
I suggest that this new way of deprecating legacy API should be approved by API review. Situation is changed.

Reply:
One way to go would be to not add new overload and live with unused parameter

Reply:
I can make the overload internal (we do have an implementation that actually overrides it and needs the options). But then other implementations of this type do not have other option that keep overriding the obsolete one.